toolkit/scripts/detect_changes.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python3
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+import argparse
+import logging
+import os
+import subprocess
+import sys
+
+from typing import Sequence
+
+logger = logging.getLogger(__name__)
+
+parser = argparse.ArgumentParser(description="Detect changes in the repository")
+parser.add_argument("--display-compare-ref", dest="display_compare_ref", action="store_true", help="Display the detected comparison ref and exit")
+parser.add_argument("--since", dest="compare_ref", required=False, help="Identifies the git ref to compare against")
+parser.add_argument("-t", "--target-ref", dest="target_ref", default="HEAD", help="Identifies the target git ref with changes")
+parser.add_argument("-u", "--include-uncommitted", dest="include_uncommitted", action="store_true", help="Include uncommitted files from working tree")
+parser.add_argument("--include-untracked", dest="include_untracked", action="store_true", help="Include untracked files in working tree")
+parser.add_argument("-v", "--verbose", dest="verbose", action="store_true", help="Enable verbose output")
+
+spec_options = parser.add_mutually_exclusive_group()
+spec_options.add_argument("-s", "--spec-names", dest="spec_names_only", action="store_true", help="Only detect changes in SPECS .spec files and print out their base names")
+spec_options.add_argument("-e", "--extended-spec-names", dest="extended_spec_names_only", action="store_true", help="Only detect changes in SPECS-EXTENDED .spec files and print out their base names")
+
+args = parser.parse_args()
+
+if args.verbose:
+    log_level = logging.DEBUG
+else:
+    log_level = logging.INFO
+
+logging.basicConfig(format="%(levelname)s: %(message)s", level=log_level)
+
+def autodetect_compare_ref(target_ref: str) -> str:
```

Comment:
The goal was for this script to be runnable in an ADO build context, a GitHub action context, and locally within a developer's environment.

If you have some specific suggestions on what APIs you have in mind, I'd be eager to look into them.

Reply:
Just looked back at this. The script *does* use GitHub PR metadata to find the comparison ref. I found the autodetection logic useful for local dev scenarios; it was otherwise ignored in the PR check.